### PR TITLE
Use user billing mocks and design

### LINKED ISSUE #4842: Implement useUserBilling mock and designs
## What is the motivation for adding / enhancing this feature?

implement `useUserBilling` composable in
- user profile
- checkout

and mock it in CT integration and boilerplate.

https://www.figma.com/file/N0Ct95cSAoODNv7zYS01ng/Storefront-UI-%7C-Design-System?node-id=4472%3A10822

The important part here is how "ADD NEW ADDRESS" button should work.

There should be a mocked list of addresses that we can push to by selecting "add new address" but it shouldn't be connected to CT. This way we will keep the data mocked and happy path working

## What are the acceptance criteria 

- [ ] useUserBilling mock for logged in user is applied in checkout in CT and boilerplate
- [ ] useUserBilling mock for logged in user is applied in user profile in CT and boilerplate
- [ ] new designs are applied for checkout

## Version of Vue Storefront

- [ ] Vue Storefront
- [x] Vue Storefront Next

### PR DESCRIPTION
**Pull request by Fifciu**
### Related Issues

closes #4842

### Breaking changes
- Added `getTaxNumber`, `getId`,`getApartmentNumber` and `isDefault` to UserShippingGetters & UserBillingGetters interfaces and their implementations
- UseUserBillingFactoryParams & UseUserShippingFactoryParams new interface. Now we have billing/shipping instead of addresses and getter to get addresses from it. We also have 2 generics instead of one inside - USER_BILLING/SHIPPING & USER_BILLING/SHIPPING_ITEM
- Implemented new UseUserBillingFactoryParams & UseUserShippingFactoryParams interfaces in core & ct & bp

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: packages/boilerplate/composables/src/composables/getters/userBillingGetters.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    if (!criteria || !Object.keys(criteria).length) {
+      return addresses;
+    }
+    const entries = Object.entries(criteria);
+    return billing.addresses.filter(
+      address => entries.every(([key, value]) => address[key] === value)
+    );
+  },
+  getDefault: billing => billing.addresses.find(address => address.isDefault),
+  getTotal: billing => billing.addresses.length,
```
Comment (filipsobol) [inline packages/boilerplate/composables/src/composables/getters/userBillingGetters.ts:15]: Please also add `getId`, `getApartment` and `getIsDefault`.

I had to add them for `useUserShipping`.

File: packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      @input="setCurrentAddress($event)"
+      class="billing__addresses"
+    >
+      <SfAddress
+        v-for="billingAddress in billingAddresses"
+        :key="billingAddress.id"
+        :name="String(billingAddress.id)"
+      >
+        <span
+          >{{ billingAddress.firstName }} {{ billingAddress.lastName }}</span
```
Comment (filipsobol) [inline packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue:14]: Please use getters to access all of these properties

File: packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      @input="setCurrentAddress($event)"
+      class="billing__addresses"
+    >
+      <SfAddress
+        v-for="billingAddress in billingAddresses"
+        :key="billingAddress.id"
+        :name="String(billingAddress.id)"
+      >
+        <span
+          >{{ billingAddress.firstName }} {{ billingAddress.lastName }}</span
```
Comment (filipsobol) [inline packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue:14]: Same as above

File: packages/commercetools/composables/src/getters/userBillingGetters.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      return addresses;
+    }
+    const entries = Object.entries(criteria);
+    return billing.addresses.filter(
+      address => entries.every(([key, value]) => address[key] === value)
+    );
+  },
+  getDefault: billing => billing.addresses.find(address => address.isDefault),
+  getTotal: billing => billing.addresses.length,
+  getPostCode: address => address ? address.zipCode : '',
```
Comment (filipsobol) [inline packages/commercetools/composables/src/getters/userBillingGetters.ts:16]: Same as above

File: packages/commercetools/theme/pages/Checkout/Payment.vue [hunk 5]
```diff
@@ -207,41 +229,128 @@ export default {
...
+        phone: address.phoneNumber
+      },
+      streetNumber: address.apartment,
+      city: address.city,
+      country: address.country,
+      state: address.state,
+      firstName: address.firstName,
+      lastName: address.lastName,
+      streetName: address.streetName,
+      postalCode: address.zipCode
```
Comment (filipsobol) [inline packages/commercetools/theme/pages/Checkout/Payment.vue:262]: Please use getters to access these properties

File: packages/commercetools/theme/pages/Checkout/Payment.vue [hunk 6]
```diff
@@ -207,41 +229,128 @@ export default {
...
+
+    onMounted(async () => {
+      if (isAuthenticated.value) {
+        await loadBilling();
+        const billingAddresses = userBillingGetters.getAddresses(billing.value);
+        if (!billingAddresses || !billingAddresses.length) {
+          return;
+        }
+        canAddNewAddress.value = false;
+        if (billingAddresses[0].isDefault) {
```
Comment (filipsobol) [inline packages/commercetools/theme/pages/Checkout/Payment.vue:289]: Getter

File: packages/core/core/src/types.ts [hunk 7]
```diff
@@ -110,18 +110,34 @@ export interface UserShippingGetters<USER_SHIPPING, USER_SHIPPING_ITEM> {
...
+  updateAddress: (address: USER_BILLING_ITEM) => Promise<void>;
   load: () => Promise<void>;
-  defaultAddress: ComputedProperty<ADDRESS>;
-  setDefault: (address: ADDRESS) => Promise<void>;
+  setDefault: (address: USER_BILLING_ITEM) => Promise<void>;
   loading: ComputedProperty<boolean>;
 }
 
+export interface UserBillingGetters<USER_BILLING, USER_BILLING_ITEM> {
+  getAddresses: (billing: USER_BILLING, criteria?: Record<string, any>) => USER_BILLING_ITEM[];
```
Comment (filipsobol) [inline packages/core/core/src/types.ts:124]: Same as above

**Review by filipsobol — CHANGES_REQUESTED** [review]

File: packages/boilerplate/composables/src/composables/getters/userBillingGetters.ts (hunk 1 above)
Comment (Fifciu) [reply packages/boilerplate/composables/src/composables/getters/userBillingGetters.ts:15]: `getApartment` - > `getApartmentNumber`

**Review by andrzejewsky — APPROVED** [review]

File: packages/boilerplate/theme/pages/MyAccount/ShippingDetails.vue [hunk 8]
```diff
@@ -150,21 +150,21 @@
         <transition-group tag="div" name="fade" class="shipping-list">
           <div
             v-for="(shipping, key) in shippingAddresses"
-            :key="shipping.streetName + shipping.apartment"
+            :key="userShippingGetters.getStreetName(shipping) + userShippingGetters.getApartmentNumber(shipping)"
```
Comment (filipsobol) [inline packages/boilerplate/theme/pages/MyAccount/ShippingDetails.vue:153]: `getId(shipping)` should give unique key

File: packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  <div>
+    <SfAddressPicker
+      :value="currentAddressId"
+      @input="setCurrentAddress($event)"
+      class="billing__addresses"
+    >
+      <SfAddress
+        v-for="billingAddress in billingAddresses"
+        :key="billingAddress.id"
+        :name="String(billingAddress.id)"
```
Comment (filipsobol) [inline packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue:11]: Use getters to access ID in `key` and `name`

File: packages/commercetools/theme/components/Checkout/UserShippingAddresses.vue [hunk 10]
```diff
@@ -11,19 +11,19 @@
         :name="String(shippingAddress.id)"
```
Comment (filipsobol) [inline packages/commercetools/theme/components/Checkout/UserShippingAddresses.vue:11]: Same as above

File: packages/commercetools/theme/pages/MyAccount/ShippingDetails.vue [hunk 11]
```diff
@@ -150,21 +150,21 @@
         <transition-group tag="div" name="fade" class="shipping-list">
           <div
             v-for="(shipping, key) in shippingAddresses"
-            :key="shipping.streetName + shipping.apartment"
+            :key="userShippingGetters.getStreetName(shipping) + userShippingGetters.getApartmentNumber(shipping)"
```
Comment (filipsobol) [inline packages/commercetools/theme/pages/MyAccount/ShippingDetails.vue:153]: `getId`

**Review by filipsobol — CHANGES_REQUESTED** [review]

**Review by filipsobol — APPROVED** [review]

File: packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    billingAddresses: {
+      type: Array,
+      required: true
+    }
+  },
+  components: {
+    SfCheckbox,
+    SfAddressPicker
+  },
+  watch: {
```
Comment (filrak) [inline packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue:67]: why not in setup?

Comment (filrak) [inline packages/commercetools/theme/components/Checkout/UserBillingAddresses.vue:67]: why not in setup?

File: packages/commercetools/theme/components/Checkout/UserShippingAddresses.vue [hunk 13]
```diff
@@ -63,17 +64,20 @@ export default {
     SfCheckbox,
     SfAddressPicker
   },
+  watch: {
+    setAsDefault (setAsDefault) {
```
Comment (filrak) [inline packages/commercetools/theme/components/Checkout/UserShippingAddresses.vue:68]: same, why not in setup?

File: packages/commercetools/theme/pages/MyAccount/ShippingDetails.vue [hunk 14]
```diff
@@ -254,19 +254,19 @@ export default {
...
-        streetName.value = shippingAddress.streetName;
-        apartment.value = shippingAddress.apartment;
-        city.value = shippingAddress.city;
-        state.value = shippingAddress.state;
-        zipCode.value = shippingAddress.zipCode;
-        country.value = shippingAddress.country;
-        phoneNumber.value = shippingAddress.phoneNumber;
-        isDefault.value = shippingAddress.isDefault;
+        const shippingAddress = userShippingGetters.getAddresses(shipping.value)[index];
+        id.value = userShippingGetters.getId(shippingAddress);
```
Comment (filrak) [inline packages/commercetools/theme/pages/MyAccount/ShippingDetails.vue:259]: can't all these be a single object? why we're making so many refs

File: packages/commercetools/theme/pages/Checkout/Payment.vue [hunk 15]
```diff
@@ -207,41 +229,129 @@ export default {
...
       chosenPaymentMethod,
       sameAsShipping,
       COUNTRIES,
-      setBillingDetails,
       setPaymentMethod,
       handleFormSubmit,
-      handleCheckSameAddress
+      handleBillingAddressSubmit,
+      handleCheckSameAddress,
+
```
Comment (filrak) [inline packages/commercetools/theme/pages/Checkout/Payment.vue:344]:
```suggestion
```

File: packages/boilerplate/composables/src/composables/useUserBilling/index.ts [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    addresses[indexToUpdate] = params.address;
+
+    if (isNewDefault) {
+      addresses.sort(sortDefaultAtTop);
+    }
+    return Promise.resolve(billing);
+  },
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  load: async (params?) => {
```
Comment (filrak) [inline packages/boilerplate/composables/src/composables/useUserBilling/index.ts:110]: we should at least provide some example data model that we would expect to fulfill getters just like in `product`

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/boilerplate/composables/src/composables/useUserBilling/index.ts [hunk 17]
```diff
@@ -0,0 +1,135 @@
+import { useUserBillingFactory, UseUserBillingFactoryParams } from '@vue-storefront/core';
+
+const addresses: any[] = [
+  {
+    id: 1,
+    email: 'john@gmail.com',
+    firstName: 'Sviatlana',
```
Comment (andrzejewsky) [inline packages/boilerplate/composables/src/composables/useUserBilling/index.ts:7]: please use some random names

File: packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  watch: {
+    setAsDefault (setAsDefault) {
+      this.localSetAsDefault = setAsDefault;
+    }
+  },
+  setup ({ setAsDefault }, { emit }) {
+    const setCurrentAddress = $event => emit('setCurrentAddress', $event);
+    const localSetAsDefault = ref(setAsDefault);
+    watch(localSetAsDefault, () => emit('changeSetAsDefault', localSetAsDefault.value));
```
Comment (andrzejewsky) [inline packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue:75]: you don't need both watchers - listen on @input or @change in the SfCheckbox

Comment (andrzejewsky) [reply packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue:75]: when it depends on the prop it's quite weird when you listen on that in the child and then update again in the parent - I think you can avoid such a watcher

**Review by andrzejewsky — CHANGES_REQUESTED** [review]

File: packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue (hunk 18 above)
Comment (Fifciu) [reply packages/boilerplate/theme/components/Checkout/UserBillingAddresses.vue:75]: Indeed

File: packages/boilerplate/composables/src/composables/useUserBilling/index.ts (hunk 16 above)
Comment (filrak) [reply packages/boilerplate/composables/src/composables/useUserBilling/index.ts:110]: in product (boilerplate) we see that `data, totals` wtc are required to be returned by `searchProducts`

**Review by filrak — APPROVED** [review]

**Pull request merged**